os/pm/pm_timer/pm_timer_activity.c

```diff
@@ -0,0 +1,180 @@
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/pm/pm.h>
+#include "pm_timer.h"
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Public Variables
+ ************************************************************************/
+
+/************************************************************************
+ * Private Functions
+ ************************************************************************/
+
+void pm_timer_expire(void) 
+{
+	pm_wakeup_timer_t *head;
+
+	while (g_pmTimer_activeList.head && ((pm_wakeup_timer_t *)g_pmTimer_activeList.head)->delay <= 0) {
+		
+    		head = (pm_wakeup_timer_t *)g_pmTimer_activeList.head;
+
+		/* Remove the pm timer from active list */
+        	sq_remfirst(&g_pmTimer_activeList);
+
+		/* As the timer is expired, give back the semaphore to unlock the thread */
+		sem_post(&head->pm_sem);
+
+		/* Make the pm timer free */
+		pm_timer_delete(head);
+	}
+}
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/****************************************************************************
+ * Name: pm_set_wakeup_timer
+ *
+ * Description:
+ *   This function is called just before sleep to start the required PM wake up
+ *   timer. It will start the first timer from the g_pmTimer_activeList with the
+ *   required delay.(delay should be positive)
+ * 
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   0 - system can go to sleep
+ *   -1 - system should not go to sleep
+ *
+ ****************************************************************************/
+
+int pm_set_wakeup_timer(void)
+{
+	/* if no timer present , do we need to wake up board ??? */
+	if (g_pmTimer_activeList.head == NULL) {
+		pmdbg("the g_pmTimer_activeList list is empty!!!\n");
+		return PM_TIMER_SUCCESS;
+	}
+
+	pm_wakeup_timer_t *curr;
+    	pm_wakeup_timer_t *prev;
+    	prev = curr = (pm_wakeup_timer_t *)g_pmTimer_activeList.head;
+
+	/* Remove the wakeup timers that are already expired. 
+	 * Although this must have already handled by pm_timer_update() */
+	while (curr->delay <= 0 && curr->next) {
+		prev = curr;
+		curr = curr->next;
+	}
+
+	if (curr->delay > 0) {
+		pmdbg("Setting timer and Board will wake up after %d millisecond\n", curr->delay);
+		up_set_pm_timer(curr->delay * 1000);
```

Comment:
This is 1 tick of TizenRT.

We can change the standard of 1 tick.

Thus, simply multiplying this by 1000 is incorrect operation.

Also, up_set_pm_timer is an API only for pm wakeup timer. I do not know why they are setting time based on hw timer in the api for pm.

Please reconsider the interval unit of up_set_pm_timer.

Reply:
multiplying by 1000 refers to changing the value from millisecond to nanosecond.